File: sdk/go/logs.go
package microsandbox

import (
	"context"
	"encoding/base64"
	"time"

	"github.com/superradcompany/microsandbox/sdk/go/internal/ffi"
)

// LogSource identifies where a persisted sandbox log entry came from.
type LogSource string

const (
	LogSourceStdout LogSource = "stdout"
	LogSourceStderr LogSource = "stderr"
	LogSourceOutput LogSource = "output"
	LogSourceSystem LogSource = "system"
)

// LogOptions filters persisted sandbox logs. Zero values read the default
// stdout and stderr sources.
type LogOptions struct {
	Tail    uint64
	Since   time.Time
	Until   time.Time
	Sources []LogSource
}

// LogEntry is one persisted sandbox log entry.
type LogEntry struct {
	Source    LogSource
	SessionID *uint64
	Timestamp time.Time
	Data      []byte
	// Cursor is an opaque resume token. Pass to LogStreamOptions.FromCursor
	// to resume a stream strictly after this entry.
	Cursor string
}

// Text returns the log payload as a string.
func (e LogEntry) Text() string { return string(e.Data) }

// LogStreamOptions configures a live log stream. The zero value reads the
// default stdout+stderr+output sources from the beginning, with follow off.
//
// Since and FromCursor are mutually exclusive — passing both rejects at the
// boundary.
type LogStreamOptions struct {
	// Sources to include. Empty means the default set
	// (stdout + stderr + output). Add LogSourceSystem for
	// runtime/kernel diagnostics.
	Sources []LogSource
	// Start at the first entry whose timestamp is >= Since.
	// Mutually exclusive with FromCursor.
	Since time.Time
	// Resume strictly after the entry whose Cursor matches this value.
	// Mutually exclusive with Since.
	FromCursor string
	// Stop emitting at the first entry whose timestamp is >= Until.
	Until time.Time
	// When true, keep the stream open past current EOF and yield new
	// entries as they are written.
	Follow bool
}

func logOptionsToFFI(opts LogOptions) ffi.LogOptions {
	out := ffi.LogOptions{
		Tail:    opts.Tail,
		Sources: make([]string, 0, len(opts.Sources)),
	}
	if !opts.Since.IsZero() {
		ms := opts.Since.UnixMilli()
		out.SinceMs = &ms
	}
	if !opts.Until.IsZero() {
		ms := opts.Until.UnixMilli()
		out.UntilMs = &ms
	}
	for _, source := range opts.Sources {
		out.Sources = append(out.Sources, string(source))
	}
	return out
}

func logEntryFromFFI(entry ffi.LogEntry) (LogEntry, error) {
	data, err := base64.StdEncoding.DecodeString(entry.DataB64)
	if err != nil {
		return LogEntry{}, err
	}
	return LogEntry{
		Source:    LogSource(entry.Source),
		SessionID: entry.SessionID,
		Timestamp: time.UnixMilli(entry.TimestampMs),
		Data:      data,
		Cursor:    entry.Cursor,
	}, nil
}

func logEntriesFromFFI(entries []ffi.LogEntry) ([]LogEntry, error) {
	out := make([]LogEntry, 0, len(entries))
	for _, entry := range entries {
		e, err := logEntryFromFFI(entry)
		if err != nil {
			return nil, err
		}
		out = append(out, e)
	}
	return out, nil
}

func logStreamOptionsToFFI(opts LogStreamOptions) ffi.LogStreamOptions {
	out := ffi.LogStreamOptions{
		Sources: make([]string, 0, len(opts.Sources)),
		Follow:  opts.Follow,
	}
	if !opts.Since.IsZero() {
		ms := opts.Since.UnixMilli()
		out.SinceMs = &ms
	}
	if opts.FromCursor != "" {
		s := opts.FromCursor
		out.FromCursor = &s
	}
	if !opts.Until.IsZero() {
		ms := opts.Until.UnixMilli()
		out.UntilMs = &ms
	}
	for _, source := range opts.Sources {
		out.Sources = append(out.Sources, string(source))
	}
	return out
}

// Logs reads persisted output for this live sandbox. It works for running and
// stopped sandboxes and does not require guest-agent protocol traffic.
func (s *Sandbox) Logs(ctx context.Context, opts LogOptions) ([]LogEntry, error) {
	entries, err := s.inner.SandboxLogs(ctx, logOptionsToFFI(opts))
	if err != nil {
		return nil, wrapFFI(err)
	}
	return logEntriesFromFFI(entries)
}

// Logs reads persisted output for this sandbox handle. It works without
// starting or connecting to the sandbox.
func (h *SandboxHandle) Logs(ctx context.Context, opts LogOptions) ([]LogEntry, error) {
	entries, err := ffi.SandboxHandleLogs(ctx, h.name, logOptionsToFFI(opts))
	if err != nil {
		return nil, wrapFFI(err)
	}
	return logEntriesFromFFI(entries)
}

// LogStreamHandle is a live log subscription. Obtain via
// Sandbox.LogStream or SandboxHandle.LogStream. Call Close to release
// Rust-side resources.
type LogStreamHandle struct {
	inner *ffi.LogStreamHandle
}

// Recv blocks until the next log entry arrives or ctx is cancelled.
// Returns nil, nil when the stream has ended (snapshot drained, Until
// reached, or a fatal stream error has already been surfaced).
func (h *LogStreamHandle) Recv(ctx context.Context) (*LogEntry, error) {
	raw, err := h.inner.Recv(ctx)
	if err != nil {
		return nil, wrapFFI(err)
	}
	if raw == nil {
		return nil, nil
	}
	entry, err := logEntryFromFFI(*raw)
	if err != nil {
		return nil, err
	}
	return &entry, nil
}

// Close stops the stream and releases Rust-side resources.
func (h *LogStreamHandle) Close() error {
	return wrapFFI(h.inner.Close())
}

// LogStream starts a streaming log subscription against a live sandbox.
// Pass LogStreamOptions{Follow: true} to keep the stream open past
// current EOF and pick up new entries as they are written. Close the
// returned handle when done.
func (s *Sandbox) LogStream(ctx context.Context, opts LogStreamOptions) (*LogStreamHandle, error) {
	h, err := s.inner.LogStream(ctx, logStreamOptionsToFFI(opts))
	if err != nil {
		return nil, wrapFFI(err)
	}
	return &LogStreamHandle{inner: h}, nil
}

// LogStream starts a streaming log subscription for this sandbox handle.
// Works without starting or connecting to the sandbox; with
// LogStreamOptions{Follow: true}, the stream picks up new entries the
// moment they land in exec.log.
func (h *SandboxHandle) LogStream(
	ctx context.Context,
	opts LogStreamOptions,
) (*LogStreamHandle, error) {
	inner, err := ffi.SandboxHandleLogStream(ctx, h.name, logStreamOptionsToFFI(opts))
	if err != nil {
		return nil, wrapFFI(err)
	}
	return &LogStreamHandle{inner: inner}, nil
}

File: sdk/node-ts/src/logs.ts
import { withMappedErrors } from "./internal/error-mapping.js";
import { mapAsyncIterable } from "./internal/async-iter.js";
import type {
  LogEntry as NapiLogEntry,
  LogOptions as NapiLogOptions,
  LogStreamOptions as NapiLogStreamOptions,
  NapiLogStream,
} from "./internal/napi.js";

/**
 * Source tag on a captured log entry.
 *
 * - `"stdout"` / `"stderr"` — the primary exec session's output in
 *   pipe mode, where the streams remain separated end to end.
 * - `"output"` — the primary exec session's merged stream when
 *   running in pty mode (pty allocation collapses stdout+stderr into
 *   a single stream at the kernel level inside the guest).
 * - `"system"` — synthetic lifecycle markers and runtime/kernel
 *   diagnostics (only emitted when explicitly requested via the
 *   `sources` option).
 */
export type LogSource = "stdout" | "stderr" | "output" | "system";

/**
 * Source filter accepted by {@link Sandbox.logs}.
 *
 * `"all"` is a convenience alias for every log source. Returned
 * entries still use concrete {@link LogSource} values.
 */
export type LogReadSource = LogSource | "all";

/**
 * One captured log entry from `exec.log`.
 *
 * Bytes are exposed via `data` as a `Uint8Array`. Use `text()` for a
 * UTF-8-lossy decode.
 */
export class LogEntry {
  /** Wall-clock capture time on the host. */
  readonly timestamp: Date;

  /** `"stdout"`, `"stderr"`, `"output"`, or `"system"`. */
  readonly source: LogSource;

  /**
   * Exec session correlation id. `null` for `system` lifecycle
   * markers, which aren't tied to a specific session. Useful when
   * grouping entries by session: `entries.filter(e => e.sessionId
   * === 42)`.
   */
  readonly sessionId: number | null;

  /** The captured chunk's bytes. */
  readonly data: Uint8Array;

  /**
   * Opaque resume token. Pass back to {@link Sandbox.logStream} as
   * {@link LogStreamOptions.fromCursor} to resume the stream
   * immediately after this entry.
   */
  readonly cursor: string;

  constructor(
    timestamp: Date,
    source: LogSource,
    sessionId: number | null,
    data: Uint8Array,
    cursor: string,
  ) {
    this.timestamp = timestamp;
    this.source = source;
    this.sessionId = sessionId;
    this.data = data;
    this.cursor = cursor;
  }

  /** UTF-8 decode of {@link data} (lossy — invalid bytes are replaced). */
  text(): string {
    return new TextDecoder("utf-8", { fatal: false }).decode(this.data);
  }
}

/**
 * Options for {@link Sandbox.logs}.
 *
 * All fields optional. Defaults: every entry, sources = `stdout +
 * stderr + output`.
 */
export interface LogReadOptions {
  /** Show only the last N entries after other filters apply. */
  tail?: number;

  /** Inclusive lower bound on entry timestamp. */
  since?: Date;

  /** Exclusive upper bound on entry timestamp. */
  until?: Date;

  /**
   * Sources to include. Defaults to `["stdout", "stderr", "output"]`
   * when omitted — i.e. all user-program output regardless of pipe
   * vs pty mode. Pass `"all"` or include `"system"` to merge
   * runtime/kernel diagnostic lines (timestamps will be a best-effort
   * approximation for unstructured kernel output).
   */
  sources?: ReadonlyArray<LogReadSource>;
}

/**
 * Options for {@link Sandbox.logStream}.
 *
 * All fields optional. `since` and `fromCursor` are mutually
 * exclusive — passing both rejects at the boundary.
 */
export interface LogStreamOptions {
  /** Same shape as {@link LogReadOptions.sources}. */
  sources?: ReadonlyArray<LogReadSource>;

  /**
   * Start at the first entry whose timestamp is `>= since`.
   * Mutually exclusive with {@link fromCursor}.
   */
  since?: Date;

  /**
   * Resume strictly after the entry whose {@link LogEntry.cursor}
   * matches this value. Mutually exclusive with {@link since}.
   */
  fromCursor?: string;

  /** Stop emitting at the first entry whose timestamp is `>= until`. */
  until?: Date;

  /**
   * When `true`, keep the stream open past current EOF and yield
   * new entries as they are written. Defaults to `false`.
   */
  follow?: boolean;
}

/**
 * An async iterable of {@link LogEntry} values.
 *
 * Use `for await...of` to drain, or call {@link recv} directly. The
 * stream ends naturally when the underlying source drains (snapshot
 * mode or `until` reached), and may end early with an error if the
 * follower falls behind the file's rotation retention window.
 */
export class LogStream
  implements AsyncIterable<LogEntry>, AsyncDisposable
{
  private readonly inner: NapiLogStream;
  private done = false;

  /** @internal */
  constructor(inner: NapiLogStream) {
    this.inner = inner;
  }

  async recv(): Promise<LogEntry | null> {
    if (this.done) return null;
    const raw = await withMappedErrors(() => this.inner.recv());
    if (raw === null) {
      this.done = true;
      return null;
    }
    return logEntryFromNapi(raw);
  }

  [Symbol.asyncIterator](): AsyncIterator<LogEntry> {
    return mapAsyncIterable(
      { recv: () => this.inner.recv() },
      logEntryFromNapi,
    )[Symbol.asyncIterator]();
  }

  async [Symbol.asyncDispose](): Promise<void> {
    this.done = true;
  }
}

/** @internal */
export function logEntryFromNapi(raw: NapiLogEntry): LogEntry {
  const source = raw.source as LogSource;
  return new LogEntry(
    new Date(raw.timestampMs),
    source,
    raw.sessionId,
    new Uint8Array(raw.data),
    raw.cursor,
  );
}

/** @internal */
export function logReadOptionsToNapi(
  opts?: LogReadOptions,
): NapiLogOptions | undefined {
  if (!opts) return undefined;
  return {
    tail: opts.tail,
    sinceMs: opts.since?.getTime(),
    untilMs: opts.until?.getTime(),
    sources: opts.sources ? Array.from(opts.sources) : undefined,
  };
}

/** @internal */
export function logStreamOptionsToNapi(
  opts?: LogStreamOptions,
): NapiLogStreamOptions | undefined {
  if (!opts) return undefined;
  return {
    sources: opts.sources ? Array.from(opts.sources) : undefined,
    sinceMs: opts.since?.getTime(),
    fromCursor: opts.fromCursor,
    untilMs: opts.until?.getTime(),
    follow: opts.follow,
  };
}
